src/DryIoc/Container.cs

```diff
@@ -11513,21 +11532,53 @@ internal object TryGetOrAddViaFactoryDelegate(int id, FactoryDelegate createValu
                 // Otherwise our own item was added and we can skip this step.
                 if (newMap == oldMap)
                 {
-                    itemRef = map.GetEntryOrDefault(id);
-                    // if someone added the item already there is a chance that it has a created value, so let's check it
-                    if (itemRef.Value != NoItem)
-                        return itemRef.Value;
+                    itemRef = map.GetSurePresentEntry(id);
+                    
+                    // If someone added the item already there is a chance it has a created value, so let's check it.
+                    // Otherwise wait until the other thread creates populates the entry.
+                    if (itemRef.Value == NoItem) 
+                    {
+#if SUPPORTS_SPIN_WAIT
+                        var spinWait = new SpinWait();
+                        while (itemRef.Value == NoItem)
+                            spinWait.SpinOnce();
+#else
+                    lock (itemRef) 
+                        while (itemRef.Value == NoItem)
+                            Monitor.Wait(itemRef);
+#endif
+                    }
+
+                    return itemRef.Value;
                 }
             }
             else 
-                itemRef = Ref.SwapAndGetNewValue(ref map, itemRef, (x, i) => x.AddOrKeepEntry(i)).GetEntryOrDefault(id);
+            {
+                var otherItemRef = Ref.SwapAndGetNewValue(ref map, itemRef, (x, i) => x.AddOrKeepEntry(i)).GetSurePresentEntry(id);
+                if (otherItemRef != itemRef) 
+                {
+#if SUPPORTS_SPIN_WAIT
+                    var spinWait = new SpinWait();
+                    while (otherItemRef.Value == NoItem)
+                        spinWait.SpinOnce();
+#else
+                    lock (itemRef) 
```

Comment:
not sure about details of lock, never read it, but if lock is somethings `have some averaging complex heuristics to spin until lock and this heuristics adds extra layers, but safes from CPU of spins in case of long locks`, so it is good and safe on average. so removing it and replacing with custom heuristics in that `spin is better in our case or write custom lock with different may be real time changed heuristics`, but give results of benchmarks seems that is of not so big help? and more risks.

Reply:
See my comment for benchmark, it is huge 10+ wing in the already highly optimized code.

And again, I almost never expect for the parallel code accessing the scope, so the majority of cases will never spin. 
Also SpinWait has an adaptive implementation as well which will start yielding and thread sleeping after cirtain amount of iterations.

Reply:
```

Reply:
what is condition for while to stop? is any chance for infinite loop?

Reply:
> `while (otherItemRef.Value == NoItem)`

The condition will be met wgen `createValue` returns. If it not returns by hanging or throwing we have the others problems to solve anyway.

Reply:
I may add the timeout here but it will just add more non-determinizm and it kind of hard to decide on the value.

Reply:
so yes, it was not about timeout, but about  otherItemRef.Value changes it state to other to allow spin to pass, so all is ok.

Reply:
> lock (itemRef)

This is a copy-paste bug, it should be lock around `otherItemRef`.